src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,184 @@
+        } else {
+            createItem(value, index);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexOutOfBoundsException(index);
+        return findByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexOutOfBoundsException(index);
+        T oldValue = findByIndex(index).item;
+        findByIndex(index).item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexOutOfBoundsException(index);
+        Node<T> currentNode = findByIndex(index);
+        unlink(currentNode);
+        return currentNode.item;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> currentNode = findByValue(object);
+        if (currentNode == null) {
+            return false;
+        }
+        unlink(currentNode);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == SIZE_EMPTY_LIST;
+    }
+
+    private Node<T> findByIndex(int index) {
+        return size() / 2 > index
+                ? searchingDown(index)
+                : searchingUp(index);
+    }
+
+    private Node<T> searchingDown(int index) {
+        Node<T> currentNode = head;
+        for (int i = 1; i <= index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode;
+    }
+
+    private Node<T> searchingUp(int index) {
+        Node<T> currentNode = tail;
+        for (int i = size() - 2; i >= index; i--) {
+            currentNode = currentNode.prev;
+        }
+        return currentNode;
+    }
+
+    private Node<T> findByValue(T value) {
+        Node<T> currentNode = head;
+        while (currentNode != null) {
+            if (equals(currentNode.item, value)) {
+                return currentNode;
+            }
+            currentNode = currentNode.next;
+        }
+        return null;
+    }
+
+    private void unlink(Node<T> node) {
+        if (node.prev != null) {
+            node.prev.next = node.next;
+        } else {
+            head = node.next;
+        }
+        if (node.next != null) {
+            node.next.prev = node.prev;
+        } else {
+            tail = node.prev;
+        }
+        size--;
+    }
+
+    private void checkIndexOutOfBoundsException(int index) {
+        if (index >= size() || index < SIZE_EMPTY_LIST) {
+            throw new IndexOutOfBoundsException("Index " + index + " out of range.");
+        }
+    }
+
+    private boolean equals(Object first, Object second) {
+        return first == second || (first != null && first.equals(second));
+    }
```

Comment:
I think you can use this logic directly, without `private boolean equals`

Reply:
ok, but...